packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,230 @@
+import { parse } from 'url'
+import { IncomingMessage, ServerResponse } from 'http'
+import { join } from 'path'
+import { mediaType } from '@hapi/accept'
+import { createReadStream, promises } from 'fs'
+import { createHash } from 'crypto'
+import Server from './next-server'
+import { fileExists } from '../../lib/file-exists'
+
+let sharp: typeof import('sharp')
+//const AVIF = 'image/avif'
+const WEBP = 'image/webp'
+const PNG = 'image/png'
+const JPEG = 'image/jpeg'
+const MIME_TYPES = [/* AVIF, */ WEBP, PNG, JPEG]
+const CACHE_VERSION = 1
+
+export async function imageOptimizer(
+  server: Server,
+  req: IncomingMessage,
+  res: ServerResponse
+) {
+  const { nextConfig, distDir } = server
+  const { sizes = [], domains = [] } = nextConfig?.experimental?.images || {}
+  const { url: reqUrl = '/', headers } = req
+  const { query } = parse(reqUrl, true)
+  const { url, w, q } = query
+  const proto = headers['x-forwarded-proto'] || 'http'
+  const host = headers['x-forwarded-host'] || headers.host
+  const mimeType = mediaType(req.headers.accept, MIME_TYPES) || ''
+
+  if (!url) {
+    res.statusCode = 400
+    res.end('"url" parameter is required')
+    return { finished: true }
+  } else if (Array.isArray(url)) {
+    res.statusCode = 400
+    res.end('"url" parameter cannot be an array')
+    return { finished: true }
+  }
+
+  let absoluteUrl: URL
+  try {
+    absoluteUrl = new URL(url)
+
+    if (
+      Array.isArray(domains) &&
+      domains.length > 0 &&
+      !domains.includes(absoluteUrl.hostname)
+    ) {
+      res.statusCode = 400
+      res.end('"url" parameter is not allowed')
+      return { finished: true }
+    }
+  } catch (_error) {
+    // url was not absolute so assuming relative url
+    try {
+      absoluteUrl = new URL(url, `${proto}://${host}`)
+    } catch (__error) {
+      res.statusCode = 400
+      res.end('"url" parameter is invalid')
+      return { finished: true }
+    }
+  }
+
+  if (!w) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) is required')
+    return { finished: true }
+  } else if (Array.isArray(w)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) cannot be an array')
+    return { finished: true }
+  }
+
+  if (!q) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) is required')
+    return { finished: true }
+  } else if (Array.isArray(q)) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) cannot be an array')
+    return { finished: true }
+  }
```

Comment:
Should we create a util to check these values to de-dupe these checks a bit?

Reply:
What would that util look like? 🧐 

These validations are doing two things: narrowing the TS type to a string and erroring when it is not a string.

We could probably do this if we create a TS assertion that throws but then we'll have several try catch.

Or we could use the type narrowing `is` return and return a single message if the parameter is missing or an array.

Another option is not throwing when the Array case is found and only use the first matching parameter via [URLSearchParams.get()](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/get).

Reply:
Ah didn't realize this was assisting with the type checks, seems good to leave it this way then 👍